lib/core/src/Cardano/Wallet/Address/Pool.hs

```diff
@@ -238,3 +247,24 @@ ensureFresh ix pool@Pool{generator,gap,addresses}
         to = toEnum $ fromEnum ix + fromIntegral gap - 1
         -- example:
         --  ix = 0 && fresh = 0 && gap = 20 `implies` [fresh .. to] = [0..19]
+
+{-------------------------------------------------------------------------------
+    Address discovery using an address pool
+-------------------------------------------------------------------------------}
+-- | Discover transactions and addresses by
+-- using an efficient query @addr -> m txs@ and an address pool.
+discover
+    :: (Enum ix, Ord addr, Monad m, Monoid txs, Eq txs)
+    => (addr -> m txs) -> Pool addr ix -> m (txs, Pool addr ix)
+discover query pool0 =
+    go mempty pool0 $ toEnum 0
+  where
+    go !txs1 !pool1 old = do
+        -- TODO: Maybe cache the `generator` in the Pool using lazy evaluation.
+        let addr = generator pool0 old
+        newtxs <- query addr
+        let pool2 = if mempty == newtxs then pool1 else update addr pool1
+            txs2  = if mempty == newtxs then txs1 else txs1 <> newtxs
```

Comment:
```suggestion
        let (pool2, txs2) = if mempty == newtxs then (pool1, txs1) else (update addr pool1, txs1 <> newtxs)
```

Reply:
😍 Something felt wrong with the previous code, but I couldn't figure out what. Now I know.